File: apps/api/services/gemini_service.py
"""
Google Gemini Service for SHELTR-AI Backend
Uses Firebase AI Logic with Vertex AI backend
"""

import os
import logging
from typing import List, Dict, Any, Optional
import google.generativeai as genai
from google.generativeai.types import HarmCategory, HarmBlockThreshold

logger = logging.getLogger(__name__)

class GeminiService:
    """Google Gemini integration service for SHELTR-AI chatbot"""
    
    def __init__(self):
        """Initialize Gemini client with SHELTR configuration"""
        try:
            # Get API key from environment
            api_key = os.getenv("GEMINI_API_KEY")
            
            if not api_key:
                logger.warning("⚠️ GEMINI_API_KEY not found in environment")
                self.available = False
                return
            
            # Configure Gemini
            genai.configure(api_key=api_key)
            
            # Safety settings (moderate blocking)
            self.safety_settings = {
                HarmCategory.HARM_CATEGORY_HARASSMENT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_HATE_SPEECH: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_SEXUALLY_EXPLICIT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
                HarmCategory.HARM_CATEGORY_DANGEROUS_CONTENT: HarmBlockThreshold.BLOCK_MEDIUM_AND_ABOVE,
            }
            
            # Generation config
            self.generation_config = {
                "temperature": 0.7,
                "top_p": 0.95,
                "top_k": 40,
                "max_output_tokens": 2000,
            }
            
            self.available = True
            logger.info("✅ Gemini service initialized successfully")
            
        except Exception as e:
            logger.error(f"❌ Failed to initialize Gemini service: {str(e)}")
            self.available = False
    
    def is_available(self) -> bool:
        """Check if Gemini service is available"""
        return self.available
    
    async def generate_chat_completion(
        self,
        messages: List[Dict[str, str]],
        model: str = "gemini-2.5-flash",
        max_tokens: int = 2000,
        temperature: float = 0.7,
        system_prompt: Optional[str] = None
    ) -> str:
        """
        Generate chat completion using Gemini
        
        Args:
            messages: List of conversation messages [{"role": "user/assistant", "content": "..."}]
            model: Gemini model to use (gemini-2.5-flash or gemini-2.5-flash-lite)
            max_tokens: Maximum tokens to generate
            temperature: Sampling temperature
            system_prompt: System instruction for the model
            
        Returns:
            Generated response text
        """
        try:
            if not self.is_available():
                raise Exception("Gemini service not available")
            
            # Create model instance
            gemini_model = genai.GenerativeModel(
                model_name=model,
                safety_settings=self.safety_settings,
                generation_config={
                    "temperature": temperature,
                    "top_p": 0.95,
                    "top_k": 40,
                    "max_output_tokens": max_tokens,
                }
            )
            
            # Convert conversation history to Gemini format
            gemini_messages = []
            
            # Add system prompt as first user message if provided
            if system_prompt:
                gemini_messages.append({
                    "role": "user",
                    "parts": [system_prompt]
                })
                gemini_messages.append({
                    "role": "model",
                    "parts": ["Understood. I'll follow these instructions."]
                })
            
            # Add conversation history
            for msg in messages:
                role = "model" if msg["role"] == "assistant" else "user"
                gemini_messages.append({
                    "role": role,
                    "parts": [msg["content"]]
                })
            
            # Start chat with history
            chat = gemini_model.start_chat(history=gemini_messages[:-1] if len(gemini_messages) > 1 else [])
            
            # Get last message to send
            last_message = gemini_messages[-1]["parts"][0] if gemini_messages else "Hello"
            
            # Generate response
            response = chat.send_message(last_message)
            
            logger.info(f"✅ Gemini ({model}) response generated successfully")
            return response.text
            
        except Exception as e:
            logger.error(f"❌ Gemini generation error: {str(e)}")
            raise Exception(f"Gemini generation failed: {str(e)}")
    
    async def generate_response(
        self,
        message: str,
        context: Dict[str, Any] = None,
        system_prompt: Optional[str] = None,
        model: str = "gemini-2.5-flash"
    ) -> str:
        """
        Generate a simple response (compatibility method)
        
        Args:
            message: User message
            context: Optional context dictionary
            system_prompt: System instruction
            model: Gemini model to use
            
        Returns:
            Generated response text
        """
        try:
            # Convert to chat format
            messages = []
            
            # Add conversation history from context if available
            if context and 'conversation_history' in context:
                for msg in context['conversation_history']:
                    # Handle different message formats
                    if isinstance(msg, dict):
                        if 'role' in msg and 'content' in msg:
                            # Standard format: {"role": "user/assistant", "content": "..."}
                            messages.append(msg)
                        elif 'user_message' in msg and 'bot_response' in msg:
                            # Orchestrator format: {"user_message": "...", "bot_response": "..."}
                            messages.append({"role": "user", "content": msg['user_message']})
                            messages.append({"role": "assistant", "content": msg['bot_response']})
            
            # Add current message
            messages.append({
                "role": "user",
                "content": message
            })
            
            return await self.generate_chat_completion(
                messages=messages,
                model=model,
                system_prompt=system_prompt
            )
            
        except Exception as e:
            logger.error(f"❌ Gemini response generation error: {str(e)}")
            raise Exception(f"Gemini response generation failed: {str(e)}")
    
    def get_available_models(self) -> List[str]:
        """Get list of available Gemini models"""
        return [
            "gemini-2.5-flash",
            "gemini-2.5-flash-lite"
        ]
    
    async def health_check(self) -> Dict[str, Any]:
        """Check Gemini service health"""
        try:
            if not self.is_available():
                return {
                    "status": "unavailable",
                    "error": "Gemini service not initialized"
                }
            
            # Try a simple generation
            test_model = genai.GenerativeModel("gemini-2.5-flash-lite")
            response = test_model.generate_content("Say 'Hello from SHELTR!'")
            
            return {
                "status": "healthy",
                "models": self.get_available_models(),
                "test_response": response.text
            }
            
        except Exception as e:
            return {
                "status": "error",
                "error": str(e)
            }

# Singleton instance
gemini_service = GeminiService()

File: apps/web/src/services/geminiService.ts
/**
 * Firebase AI Logic Service (Gemini)
 * 
 * Provides access to Gemini models via Firebase AI Logic.
 * All data stays in Google Cloud (no external API calls).
 * 
 * Models used:
 * - gemini-2.0-flash: Fast, multimodal, 1M token context (auto-updating alias)
 * - gemini-2.0-flash-lite: Faster, lighter version for simple tasks
 * 
 * References:
 * - https://firebase.google.com/docs/ai-logic
 * - https://firebase.google.com/docs/ai-logic/models
 * - https://firebase.google.com/docs/ai-logic/chat
 */

import { initializeApp, getApps } from 'firebase/app';
import { getAI, getGenerativeModel, GoogleAIBackend } from 'firebase/ai';
import type { 
  GenerativeModel, 
  ChatSession,
  GenerateContentResult,
  Content
} from 'firebase/ai';

// Firebase config (from existing Firebase setup)
const firebaseConfig = {
  apiKey: process.env.NEXT_PUBLIC_FIREBASE_API_KEY,
  authDomain: process.env.NEXT_PUBLIC_FIREBASE_AUTH_DOMAIN,
  projectId: process.env.NEXT_PUBLIC_FIREBASE_PROJECT_ID,
  storageBucket: process.env.NEXT_PUBLIC_FIREBASE_STORAGE_BUCKET,
  messagingSenderId: process.env.NEXT_PUBLIC_FIREBASE_MESSAGING_SENDER_ID,
  appId: process.env.NEXT_PUBLIC_FIREBASE_APP_ID,
  measurementId: process.env.NEXT_PUBLIC_FIREBASE_MEASUREMENT_ID,
};

/**
 * Initialize Firebase AI Logic
 * Uses Gemini Developer API backend (free tier available)
 */
class GeminiService {
  private ai: ReturnType<typeof getAI> | null = null;
  private flashModel: GenerativeModel | null = null;
  private flashLiteModel: GenerativeModel | null = null;
  private initialized = false;

  constructor() {
    this.initialize();
  }

  /**
   * Initialize Firebase and AI Logic service
   */
  private initialize() {
    try {
      // Initialize Firebase (reuse existing app if available)
      const firebaseApp = getApps().length > 0 
        ? getApps()[0] 
        : initializeApp(firebaseConfig);

      // Initialize Gemini Developer API backend
      this.ai = getAI(firebaseApp, { backend: new GoogleAIBackend() });

      // Create model instances
      // gemini-2.0-flash: Auto-updating alias to latest stable version
      this.flashModel = getGenerativeModel(this.ai, { 
        model: 'gemini-2.0-flash' 
      });

      // gemini-2.0-flash-lite: Faster, lighter version
      this.flashLiteModel = getGenerativeModel(this.ai, { 
        model: 'gemini-2.0-flash-lite' 
      });

      this.initialized = true;
      console.log('✅ Firebase AI Logic (Gemini) initialized');
    } catch (error) {
      console.error('❌ Failed to initialize Firebase AI Logic:', error);
      this.initialized = false;
    }
  }

  /**
   * Check if service is initialized and available
   */
  isAvailable(): boolean {
    return this.initialized && this.ai !== null;
  }

  /**
   * Generate text from a prompt (one-shot)
   * 
   * @param prompt - Text prompt
   * @param useLite - Use lite model for faster response (default: false)
   * @returns Generated text
   */
  async generateText(prompt: string, useLite = false): Promise<string> {
    if (!this.isAvailable()) {
      throw new Error('Gemini service not initialized');
    }

    try {
      const model = useLite ? this.flashLiteModel : this.flashModel;
      if (!model) throw new Error('Model not available');

      const result: GenerateContentResult = await model.generateContent(prompt);
      const response = result.response;
      return response.text();
    } catch (error) {
      console.error('❌ Gemini generateText error:', error);
      throw error;
    }
  }

  /**
   * Generate embeddings for text (for knowledge base search)
   * 
   * Note: Firebase AI Logic uses the same embedding model as Vertex AI
   * 
   * @param text - Text to embed
   * @returns Embedding vector
   */
  async generateEmbedding(text: string): Promise<number[]> {
    if (!this.isAvailable()) {
      throw new Error('Gemini service not initialized');
    }

    try {
      // TODO: Firebase AI Logic doesn't expose embeddings API yet
      // For now, we'll need to use Vertex AI directly or keep OpenAI
      // This is a placeholder for when the feature is available
      
      console.warn('⚠️ Embeddings not yet available via Firebase AI Logic');
      console.warn('   Continue using OpenAI for embeddings until supported');
      
      throw new Error('Embeddings not yet supported by Firebase AI Logic');
    } catch (error) {
      console.error('❌ Gemini generateEmbedding error:', error);
      throw error;
    }
  }

  /**
   * Start a chat session (multi-turn conversation)
   * 
   * @param history - Optional chat history
   * @param useLite - Use lite model for faster response (default: false)
   * @returns Chat session
   */
  startChat(history: Content[] = [], useLite = false): ChatSession {
    if (!this.isAvailable()) {
      throw new Error('Gemini service not initialized');
    }

    const model = useLite ? this.flashLiteModel : this.flashModel;
    if (!model) throw new Error('Model not available');

    return model.startChat({
      history,
      generationConfig: {
        maxOutputTokens: 8192,
        temperature: 0.7,
      },
    });
  }

  /**
   * Send a message in a chat session
   * 
   * @param session - Chat session
   * @param message - User message
   * @returns AI response
   */
  async sendChatMessage(session: ChatSession, message: string): Promise<string> {
    try {
      const result = await session.sendMessage(message);
      const response = result.response;
      return response.text();
    } catch (error) {
      console.error('❌ Gemini sendChatMessage error:', error);
      throw error;
    }
  }

  /**
   * Generate structured output (JSON)
   * 
   * @param prompt - Text prompt
   * @param schema - JSON schema for output
   * @returns Structured JSON response
   */
  async generateStructuredOutput<T = any>(
    prompt: string, 
    schema: any
  ): Promise<T> {
    if (!this.isAvailable()) {
      throw new Error('Gemini service not initialized');
    }

    try {
      if (!this.flashModel) throw new Error('Model not available');

      const result = await this.flashModel.generateContent({
        contents: [{ role: 'user', parts: [{ text: prompt }] }],
        generationConfig: {
          responseMimeType: 'application/json',
          responseSchema: schema,
        },
      });

      const response = result.response;
      const text = response.text();
      return JSON.parse(text) as T;
    } catch (error) {
      console.error('❌ Gemini generateStructuredOutput error:', error);
      throw error;
    }
  }

  /**
   * Get model information
   */
  getModelInfo() {
    return {
      flash: {
        name: 'gemini-2.0-flash',
        description: 'Fast, multimodal model with 1M token context',
        contextWindow: 1_048_576, // 1M tokens
        maxOutputTokens: 8192,
      },
      flashLite: {
        name: 'gemini-2.0-flash-lite',
        description: 'Faster, lighter version for simple tasks',
        contextWindow: 1_048_576, // 1M tokens
        maxOutputTokens: 8192,
      },
    };
  }
}

// Export singleton instance
export const geminiService = new GeminiService();

// Export types for use in other files
export type { ChatSession, Content, GenerateContentResult };

